besu/src/main/java/org/hyperledger/besu/services/TraceServiceImpl.java

```diff
@@ -88,61 +91,108 @@ public void traceBlock(final Hash hash, final BlockAwareOperationTracer tracer)
     block.ifPresent(value -> trace(value, tracer));
   }
 
-  private void trace(final Block block, final BlockAwareOperationTracer tracer) {
-    LOG.debug("Tracing block {}", block.toLogString());
-    final List<TransactionProcessingResult> results = new ArrayList<>();
+  /**
+   * Traces range of blocks
+   *
+   * @param fromBlockNumber the beginning of the range (inclusive)
+   * @param toBlockNumber the end of the range (inclusive)
+   * @param beforeTracing Function which performs an operation on a MutableWorldState before tracing
+   * @param afterTracing Function which performs an operation on a MutableWorldState after tracing
+   * @param tracer an instance of OperationTracer
+   */
+  @Override
+  public void trace(
+      final long fromBlockNumber,
+      final long toBlockNumber,
+      final Consumer<WorldState> beforeTracing,
+      final Consumer<WorldState> afterTracing,
+      final BlockAwareOperationTracer tracer) {
+    checkArgument(tracer != null);
+    LOG.debug("Tracing from block {} to block {}", fromBlockNumber, toBlockNumber);
+    final Blockchain blockchain = blockchainQueries.getBlockchain();
+    final List<Block> blocks =
+        LongStream.rangeClosed(fromBlockNumber, toBlockNumber)
+            .mapToObj(
+                number ->
+                    blockchain
+                        .getBlockByNumber(number)
+                        .orElseThrow(() -> new RuntimeException("Block not found " + number)))
+            .toList();
     Tracer.processTracing(
         blockchainQueries,
-        block.getHash(),
+        blocks.get(0).getHash(),
```

Comment:
This looks clean and succinct.  But I think we at least need a unit test that builds block 1, 2, 3 where the transactions of block 2 and 3 would fail to execute against the parent worldstate of block 1, but succeed when executed against their correct parent worldstates.  

When I first looked at the implementation it looked like we would be executing all of the blocks against the (unmutated) parent worldstate of block[0]

Reply:
good catch I added tests and found a bug. It's working correctly now